### PR TITLE
[DNNL][INT8][FP32] MatMul

### PR DESCRIPTION
**Pull request by Sand3r-**
This PR introduces:
* FP32 and INT8 implementations of DNNL MatMul operator
* UTs to test functionality

### CODE REVIEW

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+  template <typename T>
+  dnnl::memory CreateMemory(const memory::dims& dims,
+                            const memory::dims& strides, const T* data) {
+    auto md = memory::desc(dims, MKLDNNGetDataType<T>(), strides);
+    return dnnl::memory(md, engine_, to_void_cast(data));
+  }
+
+  MatMulDims GetMatmulDims(const ExecutionContext& ctx) {
+    auto dim_x = math::CreateMatrixDescriptor(
```
Comment (sfraczek) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: I would just suggest that you could consider renaming dim_x -> mat_dim_x like in matmul_op.cc InferShape https://github.com/PaddlePaddle/Paddle/pull/23395/files#diff-126803a76a1b0cbf6198f30e23058f8dR336.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+#include "paddle/fluid/operators/math/blas.h"
+#include "paddle/fluid/platform/mkldnn_helper.h"
+
+namespace paddle {
+namespace operators {
+
+using dnnl::memory;
+using dnnl::primitive;
+using platform::to_void_cast;
+using Tensor = framework::Tensor;
```
Comment (sfraczek) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: This one stands out in the middle. What do you think about for example moving it to bottom of block or something?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        self.set_attributes()
+        self.attrs['use_mkldnn'] = True
+
+        self.inputs = {'X': self.x, 'Y': self.y}
+        self.outputs = {'Out': self.out}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad_normal(self):
```
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: skip_check_grad_ci is recommended in such case. Like following
https://github.com/PaddlePaddle/Paddle/pull/21985/files

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc (hunk 1 above)
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Done.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc (hunk 2 above)
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Done.

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py (hunk 3 above)
Comment (Sand3r-) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: Done.

**Comment by lidanqing-vv** [comment]
@luotao1  Hi, please give an approval about the UT `@skip_check_grad_ci`. This is blocking the CI to pass. Thanks

**Comment by Sand3r-** [comment]
@luotao1 

PR-CI-Coverage failed because it cannot find error_codes when building shell.cc. I didn't touch that file. What should I do?
```
[2020-04-03 21:24:05]  [  3%] Linking CXX static library libfs.a
[2020-04-03 21:24:05]  In file included from /paddle/paddle/fluid/platform/enforce.h:41:0,
[2020-04-03 21:24:05]                   from /paddle/paddle/fluid/framework/io/shell.cc:16:
[2020-04-03 21:24:05]  /paddle/paddle/fluid/platform/errors.h:23:50: fatal error: paddle/fluid/platform/error_codes.pb.h: No such file or directory
[2020-04-03 21:24:05]  #include "paddle/fluid/platform/error_codes.pb.h"
[2020-04-03 21:24:05]                                                    ^
[2020-04-03 21:24:05]  compilation terminated.
```

**Comment by luotao1** [comment]
The fixed PR is #23467, please rerun after it is merged.

**Comment by sfraczek** [comment]
Can you add matmul to `--quantized_ops` in `python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py`  and   `QAT2_NLP_QUANTIZED_OPS` in `python/paddle/fluid/contrib/slim/tests/CMakeLists.txt`?

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+using framework::DataLayout;
+using platform::GetMKLDNNFormat;
+using platform::MKLDNNGetDataType;
+using platform::MKLDNNDeviceContext;
+using framework::ExecutionContext;
+using Tensor = framework::Tensor;
+
+// Get row matrix shape from a vector shape. If the rank of x_dim > 1, the
+// original x_dim is returned.
+static framework::DDim RowMatrixFromVector(const framework::DDim& x_dim) {
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: The name is a bit misleading, as the function is about the shape, not about the matrix.
Same below

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+import unittest, os
+import numpy as np
+from paddle.fluid.tests.unittests.op_test import OpTest, skip_check_grad_ci
+
+
+@skip_check_grad_ci(reason="DNNL's MatMul doesn't implemend grad kernel.")
+class TestDnnlMatMulOp(OpTest):
+    def generate_data(self):
+        self.x = np.random.random((1, 2, 2)).astype("float32")
```
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: Aren't the dimensions too small for CI? I remember a requirement that tensors should have at least 100 elements.

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        self.set_attributes()
+        self.attrs['use_mkldnn'] = True
+
+        self.inputs = {'X': self.x, 'Y': self.y}
+        self.outputs = {'Out': self.out}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad_normal(self):
```
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: You can drop the grad checking methods after adding `@skip_check_grad_ci`.

**Review by wojtuss — COMMENTED** [review]
I find this PR very concise and elegant. My comments are very minor and, as the CI is not stable and the queue is very long, please, do not consider them as requests for changes. They can be addressed later in a refactoring PR if necessary.

**Comment by Sand3r-** [comment]
> Can you add matmul to `--quantized_ops` in `python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py` and `QAT2_NLP_QUANTIZED_OPS` in `python/paddle/fluid/contrib/slim/tests/CMakeLists.txt`?

This PR doesn't introduce any changes to cpu quantise pass, so that wouldn't work, nope

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py (hunk 5 above)
Comment (lidanqing-vv) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: @Sand3r-  https://github.com/PaddlePaddle/Paddle/wiki/OP-Test-Input-Shape-Requirements-(English-Version) The requirements wojtek mentioned could be found here. @skip_check_grad is here too

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc (hunk 4 above)
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: What other name do you suggest instead? I think it conveys the meaning that it accepts a vector of size N and converts it into a matrix of size 1 x N. Similarly for the other function.

Comment (wojtuss) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Maybe `RowMatrixDimsFromVector`, or `VectorDims2RowMatrixDims`?
It is only a suggestion.

Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: But it's the same function as already exists here:
https://github.com/PaddlePaddle/Paddle/pull/23395/files/471e894b65b0cf2db48de2ee2b0c330d3cbd7815#diff-126803a76a1b0cbf6198f30e23058f8dL39

**Review by wojtuss — APPROVED** [review]
LGTM

**Comment by Sand3r-** [comment]
@luotao1 
All builds have passed (apart from the one that needs approval) and internal review is already finished. Could you please start your review?

**Review by sfraczek — APPROVED** [review]
LGTM

File: paddle/fluid/operators/matmul_op.cc [hunk 7]
```diff
@@ -426,6 +444,30 @@ class MatMulOpMaker : public framework::OpProtoAndCheckerMaker {
...
+    AddAttr<float>("Scale_y",
+                   "(float, default 1.0f), The quantize scale of Y tensor")
+        .SetDefault(1.0f);
+    AddAttr<float>("Scale_out",
+                   "(float, default 1.0f), The quantize scale of output data")
+        .SetDefault(1.0f);
+    AddAttr<bool>("force_fp32_output",
+                  "(bool, default false) Force INT8 kernel output FP32, only "
+                  "used in MKL-DNN INT8")
+        .SetDefault(false);
```
Comment (luotao1) [inline paddle/fluid/operators/matmul_op.cc:470]: Are these attributes all needed?

Comment (Sand3r-) [reply paddle/fluid/operators/matmul_op.cc:470]: Yes, all int8 compute operators need these attributes. 
Conv2d: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/conv_op.cc#L307-L326
FC: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/fc_op.cc#L145-L165

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by lidanqing-vv** [comment]
@luotao1  Can you please help restart PR_CI_CPU_Py2 (Compile)? As you already approved, but this CI was not restarted.

**Pull request merged**